### PR TITLE
Roundabouts into ferries

### PR DESCRIPTION
**Pull request by MoKob**
# Issue

Resolves https://github.com/Project-OSRM/osrm-backend/issues/3762

## Tasklist
 - [x] Fix Roundabout into ferry handling
       - see discussion over at ticket
 - [x] update relevant [Wiki pages](https://github.com/Project-OSRM/osrm-backend/wiki)
 - [x] add regression / cucumber cases (see docs/testing.md)
 - [x] review
 - [x] adjust for comments

### CODE REVIEW

File: include/engine/guidance/route_step.hpp [hunk 1]
```diff
@@ -144,6 +146,14 @@ inline RouteStep &RouteStep::AddInFront(const RouteStep &preceeding_step)
     return *this;
 }
 
+// checks whether we can actually merge two route steps:
+// They neeed to have the same mode, because a route step
+// can only have one mode.
+inline bool RouteStep::IsCompatible(const RouteStep &other) const
+{
+    return mode == other.mode;
+}
```
Comment (daniel-j-h) [inline include/engine/guidance/route_step.hpp]: Hm I'm a bit out of the loop here but I thought we had this exact function already.
Can't see it on the route step, so this should be fine for sure.

File: include/engine/guidance/route_step.hpp [hunk 2]
```diff
@@ -144,6 +146,14 @@ inline RouteStep &RouteStep::AddInFront(const RouteStep &preceeding_step)
     return *this;
 }
 
+// checks whether we can actually merge two route steps:
+// They neeed to have the same mode, because a route step
+// can only have one mode.
```
Comment (daniel-j-h) [inline include/engine/guidance/route_step.hpp]: Maybe add the comments to the decl. and not the impl.

File: include/storage/shared_monitor.hpp [hunk 3]
```diff
@@ -12,7 +12,7 @@
 #include <boost/interprocess/sync/scoped_lock.hpp>
 
 #if defined(__linux__)
-#define USE_BOOST_INTERPROCESS_CONDITION 1
+//#define USE_BOOST_INTERPROCESS_CONDITION 1
```
Comment (daniel-j-h) [inline include/storage/shared_monitor.hpp]: Do you want to revert this before merging?

File: src/engine/guidance/post_processing.cpp [hunk 4]
```diff
@@ -160,6 +162,21 @@ void closeOffRoundabout(const bool on_roundabout,
...
+    if (step_index > 1)
+    {
+        // In case the step with the roundabout exit instruction cannot be merged with the
+        // previous step we change the instruction to a normal turn
+        if (!steps[step_index].IsCompatible(steps[step_index - 1]))
+        {
+            BOOST_ASSERT(leavesRoundabout(steps[step_index].maneuver.instruction));
+            steps[step_index-1].maneuver.instruction = steps[step_index].maneuver.instruction;
+            if (!entersRoundabout(steps[step_index-1].maneuver.instruction))
+                steps[step_index-1].maneuver.exit = steps[step_index].maneuver.exit;
```
Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: can you alias steps[step_index - 1] for readability

File: src/engine/guidance/post_processing.cpp [hunk 5]
```diff
@@ -160,6 +162,21 @@ void closeOffRoundabout(const bool on_roundabout,
...
+            steps[step_index].maneuver.instruction.type = TurnType::Notification;
```
Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: Notification to signal a change in travel mode?

Comment (MoKob) [inline src/engine/guidance/post_processing.cpp]: That is our designed behaviour.

File: include/engine/guidance/route_step.hpp (hunk 1 above)
Comment (TheMarex) [inline include/engine/guidance/route_step.hpp]: We have a similar concept for graph compression, I didn't find anything explicit for guidance.

Comment (MoKob) [inline include/engine/guidance/route_step.hpp]: https://github.com/Project-OSRM/osrm-backend/blob/56bcb491d324f6258cff74dd2227e4350cc51460/include/engine/guidance/collapsing_utility.hpp#L109

**Pull request merged**